torch_geometric/data/feature_store.py

```diff
@@ -0,0 +1,227 @@
+
+This particular feature store abstraction makes a few key assumptions:
+    * The features we care about storing are all associated with some sort of
+        `index`; explicitly for PyG the the index of the node in the graph (or
+            the heterogeneous component of the graph it resides in).
+    * A feature can uniquely be identified from (a) its index and (b) any other
+        associated attributes specified in :obj:`TensorAttr`.
+
+It is the job of a feature store implementor class to handle these assumptions
+properly. For example, a simple in-memory feature store implementation may
+concatenate all metadata values with a feature index and use this as a unique
+index in a KV store. More complicated implementations may choose to partition
+features in interesting manners based on the provided metadata.
+
+Major TODOs for future implementation:
+* Async `put` and `get` functionality
+"""
+from abc import abstractmethod
+from collections.abc import MutableMapping
+from dataclasses import dataclass
+from typing import Any, Optional
+
+import numpy as np
+import torch
+
+from torch_geometric.typing import FeatureTensorType
+from torch_geometric.utils.mixin import CastMixin
+
+
+@dataclass
+class TensorAttr(CastMixin):
+    r"""Defines the attributes of a :obj:`FeatureStore` tensor."""
+
+    # The node indices the rows of the tensor correspond to
+    index: Optional[FeatureTensorType] = None
+
+    # The type of the feature tensor (may be used if there are multiple
+    # different feature tensors for the same node index)
+    tensor_type: Optional[str] = None
+
+    # The type of the nodes that the tensor corresponds to (may be used for
+    # hetereogeneous graphs)
+    node_type: Optional[str] = None
+
+    # The type of the graph that the nodes correspond to (may be used if a
+    # feature store supports multiple graphs)
+    graph_type: Optional[str] = None
+
+
+class AttrView:
+    r"""A view of a :obj:`FeatureStore` that is obtained from an incomplete
+    specification of attributes. This view stores a reference to the
+    originating feature store as well as a :obj:`TensorAttr` object that
+    represents the view's (incomplete) state.
+
+    As a result, store[TensorAttr(...)].tensor_type[idx] allows for indexing
+    into the store.
+    """
+    _store: 'FeatureStore'
+    attr: TensorAttr
+
+    def __init__(self, store, attr):
+        self._store = store
+        self.attr = attr
+
+    def __getattr__(self, tensor_type):
+        r"""Supports attr_view.attr"""
+        self.attr.tensor_type = tensor_type
+        return self
+
+    def __getitem__(self, index: FeatureTensorType):
+        r"""Supports attr_view.attr[idx]"""
+        self.attr.index = index
+        return self._store.get_tensor(self.attr)
+
+    def __repr__(self) -> str:
+        return f'AttrView(store={self._store}, attr={self.attr})'
+
+
+class FeatureStore(MutableMapping):
+    def __init__(self, backend: Any):
```

Comment:
`backend` doesn't seem to be used anywhere. Why not drop`__init__` completely here?